rcl_yaml_param_parser/src/parser.c

```diff
@@ -0,0 +1,1384 @@
+            allocator.deallocate(ret_val, allocator.state);
+            return RCL_RET_ERROR;
+          }
+        }
+        res = add_val_to_double_arr(param_value->double_array_value, ret_val, allocator);
+        if (RCL_RET_OK != res) {
+          if (NULL != ret_val) {
+            allocator.deallocate(ret_val, allocator.state);
+          }
+          return res;
+        }
+      }
+      break;
+    case DATA_TYPE_STRING:
+      if (false == is_seq) {
+        param_value->string_value = (char *)ret_val;
+      } else {
+        if (DATA_TYPE_UNKNOWN == *seq_data_type) {
+          *seq_data_type = val_type;
+          param_value->string_array_value =
+            allocator.zero_allocate(1U, sizeof(rcutils_string_array_t), allocator.state);
+          if (NULL == param_value->string_array_value) {
+            RCUTILS_SAFE_FWRITE_TO_STDERR("Error allocating mem");
+            return RCL_RET_BAD_ALLOC;
+          }
+        } else {
+          if (*seq_data_type != val_type) {
+            RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(allocator, "Sequence should be of same"
+              " type. Value type 'string' do not belong at line_num %d", line_num);
+            allocator.deallocate(ret_val, allocator.state);
+            return RCL_RET_ERROR;
+          }
+        }
+        res = add_val_to_string_arr(param_value->string_array_value, ret_val, allocator);
+        if (RCL_RET_OK != res) {
+          if (NULL != ret_val) {
+            allocator.deallocate(ret_val, allocator.state);
+          }
+          return res;
+        }
+      }
+      break;
+    default:
+      RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(allocator, "Unknown data type of value"
+        " %s at line %d", value, line_num);
+      res = RCL_RET_ERROR;
+      break;
+  }
+  return res;
+}
+
+///
+/// Parse the key part of the <key:value> pair
+///
+static rcl_ret_t parse_key(
+  const yaml_event_t event,
+  uint32_t * map_level,
+  bool * is_new_map,
+  namespace_tracker_t * ns_tracker,
+  rcl_params_t * params_st,
+  const rcl_allocator_t allocator)
+{
+  int32_t res = RCL_RET_OK;
+  const size_t val_size = event.data.scalar.length;
+  const char * value = (char *)event.data.scalar.value;
+  const uint32_t line_num = ((uint32_t)(event.start_mark.line) + 1U);
+  uint32_t num_nodes;
+  uint32_t node_idx = 0U;
+
+  if ((NULL == map_level) || (NULL == params_st)) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  if (val_size > MAX_STRING_SIZE) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(allocator, "Scalar value at line %d"
+      " is bigger than %d bytes", line_num, MAX_STRING_SIZE);
+    return RCL_RET_ERROR;
+  } else {
+    if (0U == val_size) {
+      RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(allocator, "No key at line %d", line_num);
+      return RCL_RET_ERROR;
+    }
+  }
+
+  if (NULL == value) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  num_nodes = params_st->num_nodes;  // New node index
```

Comment:
On windows I see a warning about a possible loss of data here because of the conversion from size_t to `uint32_t`

Reply:
Make sense